test/bats/helpers.bash

```diff
@@ -70,6 +70,7 @@ get_ca_cert() {
   if [ $(kubectl get secret -n gatekeeper-system gatekeeper-webhook-server-cert -o jsonpath='{.data.ca\.crt}' | wc -w) -eq 0 ]; then
     return 1
   fi
+  sleep 10s
```

Comment:
@ritazh @maxsmythe  - I personally did not want to do this. But the testing was really flaky.
I triggered the testing 4 times and every single time it failed with timeout.

 I reduced the wait time to 2 to see what the issue was. And the portforwarding had issues.

```#  in test file test/bats/test.bats, line 37)
#   `assert_success' failed
# Forwarding from 127.0.0.1:8443 -> 8443
# Forwarding from [::1]:8443 -> 8443
# Handling connection for 8443
# E0929 15:24:30.342505   14834 portforward.go:400] an error occurred forwarding 8443 -> 8443: error forwarding port 8443 to pod 23cabab5c248015fc95dd161aa4c703c38c3abbc41828d84c021a44d2bac9403, uid : failed to execute portforward in network namespace "/var/run/netns/cni-4ba472a7-6312-4d2e-8e31-c0b42f519bf2": failed to dial 8443: dial tcp4 127.0.0.1:8443: connect: connection refused
# expected: 0
# actual: 1
```
And it looked like webhook server itself hasnt started listening on 8443.
It felt like we are relying on the presence of this secret for webhook to determine if the port 8443 has been listened on.

So I added a delay to provide some time for the webhook server to start listening on 8443 before we port forward.
To me, this should be handled by the readiness probe. I did not really dig into the implementation of the health checker. 

Thoughts ?

Reply:
This is a good find! We are indeed just assuming port forwarding will work.

The difficulty here is that port-forwarding needs to happen in the background, so it's a bit harder to test for success. We could maybe use `netstat`:

```bash
port_forwarding_enabled() {
    netstat -ltnp | grep 127.0.0.1:8443 | grep LISTEN
}
```

```bash
port_forwarding_pid() {
    netstat -ltnp | grep 127.0.0.1:8443 | awk '{print $NF}' | awk -F/ '{print $1}'
}
```

```bash
wait_for_process_failure() {
    wait_time="$1"
    sleep_time="$2"
    cmd="$3"
    while [ "$wait_time" -gt 0 ]; do
        if eval "$cmd"; then
            sleep "$sleep_time"
            wait_time=$((wait_time - sleep_time))
        else
            return 0
        fi
  done
  return 1
}
```

```bash
disable_port_forwarding() {
    if port_forwarding_enabled; then
        kill "$(port_forwarding_pid)"
        wait_for_process_failure  5 1 "port_forwarding_enabled"
    fi
}
```

```bash
enable_port_forwarding() {
    echo "> enabling port forwarding" >&3
    disable_port_forwarding
    kubectl port-forward -n gatekeeper-system deployment/gatekeeper-controller-manager 8443:8443 &
    wait_for_process 5 1 "port_forwarding_enabled"
}
```

We should then wrap the call to `enable_port_forwarding` in a retry loop.

The `> &3` above allows us to bypass the stdout capture that bats normally does so we can actually see what we are hanging on.

Note that I haven't tested this code, so there may be some bugs. We are also enabling port forwarding to the metrics ports, so the above code may need to be generalized.

Reply:
Definitely worth trying. Do you think this PR can go ahead ? 
I have referenced this conversation to the issue that is tracking the e2e timeout. https://github.com/open-policy-agent/gatekeeper/issues/848
I can spend some time this week on this and raise a PR :)

Reply:
@maxsmythe  - Another idea would be to run a pod with curl in the kind cluster and exec curl command in there using the wait logic. The advantage would be that we would not have to deal with the network and port conflict kind of scenarios in future.

I think portforwarding as a background process and terminating the pid is too complicated for a testcase validating if the kubernetes service endpoint is responding from within the cluster. 

During timeouts, the teardown or clean up functions also does not really get executed.

Reply:
Raised a PR : https://github.com/open-policy-agent/gatekeeper/pull/861

Reply:
No issues adding the sleep, it at least mitigates the problem.